src/libraries/System.Private.CoreLib/src/ILLink/ILLink.Substitutions.Shared.xml

```diff
@@ -1,5 +1,9 @@
 <linker>
   <assembly fullname="System.Private.CoreLib">
+    <type fullname="System.Diagnostics.Tracing.EventSource" feature="System.Diagnostics.Tracing.EventSource.IsSupported" featurevalue="false">
+      <method signature="System.Boolean get_IsSupported()" body="stub" value="false" />
+      <method signature="System.Boolean IsEnabled()" body="stub" value="false" />
```

Comment:
Why not stub out the other two IsEnabled overloads as well?

Reply:
The ILLinker doesn't support inlining methods with parameters.

https://github.com/mono/linker/blob/b68b74fa3814e49db4b4743014f0adc468e45700/src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs#L188-L189

Reply:
> The ILLinker doesn't support inlining methods with parameters

I'm confused; based on @marek-safar's comments in https://github.com/dotnet/runtime/pull/37744#discussion_r440096144, I thought the linker didn't do any inlining at all.  So... it does inline?

Reply:
It depends on how you define the word "inline".

What the linker does when it sees a method that returns a constant value, it takes that constant value and inserts it into the callsite, so it can trim unused branches. However, it also leaves the method call in the callsite as well. You can see this above in the 3rd `IsEnabled` overload. The original code is:

```C#
        public bool IsEnabled(EventLevel level, EventKeywords keywords, EventChannel channel)
        {
            if (!IsEnabled())
                return false;

            if (!IsEnabledCommon(m_eventSourceEnabled, m_level, m_matchAnyKeyword, level, keywords, channel))
                return false;
            return true;
        }
```

And what is left after trimming when `IsEnabled()` is stubbed to be `return false;` is:

```C#
        public bool IsEnabled(EventLevel level, EventKeywords keywords, EventChannel channel)
        {
            this.IsEnabled();
            return false;
        }
```

I don't know of a better word to use for this behavior. The method that does this in the linker is named [`TryInlineBodyDependencies`](https://github.com/mono/linker/blob/b68b74fa3814e49db4b4743014f0adc468e45700/src/linker/Linker.Steps/RemoveUnreachableBlocksStep.cs#L160).

Reply:
Thanks.  Why does the linker prevent changing the body of this method to be a stub?  I'd understand it if it had out arguments, but why can't specifying:
```XML
<method signature="System.Boolean IsEnabled(EventLevel, EventKeywords, EventChannel)" body="stub" value="false" />
```
cause the method to become:
```
public bool IsEnabled(EventLevel level, EventKeywords keywords, EventChannel channel) => false;
```
?

Reply:
Linker tries to be super careful not to break things - so it has lot of limitations put in place just because it was not needed. I think it would be fine to lift this restriction (similar to lifting the instance method restriction which made this available - only propagate the constant return value on methods with explicit substitutions).

@marek-safar @MichalStrehovsky - any objections to doing that? That is, can you think of a case where doing so would have unintended side-effects? (other than changing the method body specified in substitutions)

Reply:
We shouldn't allow the "inlining"/stubbing if any of the parameters are byrefs (because it could leave us with an uninitialized variable when combined with SkipLocalsInit), but it should be fine otherwise.

Providing means to avoid users having to `if (IsEnabled() && IsEnabled(EventLevel.Warning, Keywords.AsyncMethod))` to get good trimming seems like goodness. (EventSource is a public API so whatever we do here is the customer experience.)

Reply:
Yeah, we could implement this extension if it's needed.

Reply:
I've logged https://github.com/mono/linker/issues/1288 for this. I'll update the Substitutions.xml to stub out all 3 overloads of `EventSource::IsEnabled`, so it will "just work" when the linker enables this substitution.